Retest for realtime effects

Paul-Licameli opened the pull request:
Resolves: #2928

Bump the version number of the plugin registry config file; use that to decide when a rescan of all built-in effects is needed
at startup for changes of realtime capability of built-in effects.

This will do the right thing either upgrading to 3.2 or later (so that Reverb is considered realtime), or downgrading from a
future Audacity version, with a further changed plugin registry version, to an earlier Audacity, but not before 3.2.

The problem of downgrading from 3.2 to 3.1 may go slightly wrong -- then 3.1 will mistreate Reverb as a realtime effect with
a non-modal dialog.

An automatic fix for this further problem could be devised, but the question is whether it's worth the trouble?  If anyone
really meets this problem, they can be told to delete pluginresgistry.cfg or edit certain lines out of it.

Update:  added three commits that do make the extra effort to avoid surprises in behavior of the Reverb effect, in case of a downgrade from 3.2 back to 3.1.

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

crsib reviewed (dismissed):
Using `pluginregistryversion` as a marker for the rescan should at least be documented well somewhere on the developers Wiki. I don't have a better suggestion though.

I will be happy if serialization is done in a less hacky way, but that is not strong enough reason to block this PR

Paul-Licameli commented on libraries/lib-components/EffectInterface.h:
```diff
@@ -193,8 +193,23 @@ class COMPONENTS_API EffectDefinitionInterface  /* not final */
+   //! In which versions of Audacity was an effect realtime capable?
+   enum class RealtimeSince : unsigned {
+      Never,
+      Since_3_2,
```
The point of all my hackery is to let a downgrade from 3.2 to 3.1 cause Reverb to be treated correctly in 3.1 as not realtime capable.  The rewrite won't work.

Maybe we should make a rule that we don't care about downgrades of more than one minor version -- which would mean that 3.3 and later might make this simplification.

Paul-Licameli commented on libraries/lib-module-manager/PluginDescriptor.cpp:
```diff
@@ -194,11 +194,50 @@ void PluginDescriptor::SetEffectLegacy(bool legacy)
    mEffectRealtime = realtime;
 }
 
+static constexpr auto Since_3_2_string = "00";
```
See previous:  downgrade to 3.1 would not work.

Paul-Licameli commented on libraries/lib-module-manager/PluginInterface.h:
```diff
@@ -80,6 +80,29 @@ using ConfigConstReference =
+   return regver1 == regver2;
+}
+
+// Compare registry versions lexicographically as std::wstring
```
I was doubtful that wxString comparison might have some strange locale sensitivity?

But no, I decided to be less lazy in my rewrite.  No more std:: strings.  Instead parse the numbers form the string and compare the number sequences lexicographically.  Do it right in case any of the numbers reach double digits.

Paul-Licameli commented on libraries/lib-module-manager/PluginManager.h:
```diff
@@ -222,4 +227,7 @@ class MODULE_MANAGER_API PluginManager final : public PluginManagerInterface
 #define NYQUIST_PROMPT_NAME XO("Nyquist Prompt")
 
+// Latest version of the plugin registry config
+#define REGVERCUR wxString(wxT("1.2"))
```
yeah

Paul-Licameli commented:
@crsib @pietro68 Please review again.

I rewrote the comparison of versions more properly, and I fixed an unintended change of behavior for report of realtime capability from AudioUnit effects.
